File: ts-sdk/src/voice.ts
/**
 * Voice capability API
 */
import { getAgentsWsUrlFromProductBase } from './config.js';
import { requestWithRetry, responseAPIError, type HTTPClient } from './http.js';
import { VoiceSession, type VoiceChunkEvent } from './voice-session.js';
import {
  VoiceTranscriptionSession,
  type TranscriptionStartOptions,
  type VoiceTranscriptionEvent,
} from './voice-transcription-session.js';

export {
  VoiceSession,
  VoiceStreamError,
  type CloneSpeakOptions,
  type DesignSpeakOptions,
  type SpeakOptions,
  type VoiceChunkEvent,
  type VoiceSessionOptions,
  type VoiceSessionState,
} from './voice-session.js';

export {
  VoiceTranscriptionSession,
  type TranscriptionStartOptions,
  type VoiceTranscriptDeltaEvent,
  type VoiceTranscriptFinalEvent,
  type VoiceTranscriptionAckEvent,
  type VoiceTranscriptionEvent,
  type VoiceTranscriptionSessionOptions,
  type VoiceTranscriptionSessionState,
} from './voice-transcription-session.js';

export interface TTSOptions {
  text: string;
  voice?: string;
  language?: string;
  responseFormat?: string;
  signal?: AbortSignal;
}

export interface CloneOptions {
  text: string;
  refAudio: Uint8Array | ArrayBuffer;
  language?: string;
  xVectorOnly?: boolean;
  responseFormat?: string;
}

export interface DesignOptions {
  text: string;
  description: string;
  language?: string;
  responseFormat?: string;
}

export interface TranscribeOptions {
  audio: Uint8Array | ArrayBuffer | Blob;
  filename?: string;
  contentType?: string;
  language?: string;
  model?: string;
  responseFormat?: string;
  prompt?: string;
  signal?: AbortSignal;
}

export interface TranscriptionResult {
  text: string;
  [key: string]: unknown;
}

export interface TranscribeStreamOptions {
  audio: Uint8Array | ArrayBuffer | string;
  language?: string;
  model?: string;
  responseFormat?: string;
  prompt?: string;
  base64?: boolean;
  timeoutMs?: number;
}

function encodeBase64(bytes: Uint8Array | ArrayBuffer): string {
  const data = bytes instanceof Uint8Array ? bytes : new Uint8Array(bytes);
  if (typeof Buffer !== 'undefined') {
    return Buffer.from(data).toString('base64');
  }
  // Browser fallback: btoa over chunked binary string (avoids arg-limit blowups)
  let binary = '';
  const CHUNK = 0x8000;
  for (let i = 0; i < data.length; i += CHUNK) {
    binary += String.fromCharCode(...data.subarray(i, i + CHUNK));
  }
  return btoa(binary);
}

export class VoiceAPI {
  constructor(private http: HTTPClient) {}

  async tts(options: TTSOptions): Promise<Uint8Array> {
    return this.http.postBytes('/voice/tts', {
      text: options.text,
      voice: options.voice ?? 'serena',
      language: options.language ?? 'auto',
      response_format: options.responseFormat ?? 'mp3',
    }, options.signal ? { signal: options.signal } : {});
  }

  async clone(options: CloneOptions): Promise<Uint8Array> {
    return this.http.postBytes('/voice/clone', {
      text: options.text,
      ref_audio_base64: encodeBase64(options.refAudio),
      language: options.language ?? 'auto',
      x_vector_only: options.xVectorOnly ?? true,
      response_format: options.responseFormat ?? 'mp3',
    });
  }

  async design(options: DesignOptions): Promise<Uint8Array> {
    return this.http.postBytes('/voice/design', {
      text: options.text,
      instruct: options.description,
      language: options.language ?? 'auto',
      response_format: options.responseFormat ?? 'mp3',
    });
  }

  async transcribe(options: TranscribeOptions): Promise<TranscriptionResult> {
    const formData = new FormData();
    const filename = options.filename ?? 'audio';
    const audio = options.audio instanceof Blob
      ? options.audio
      : new Blob([options.audio as unknown as NonNullable<ConstructorParameters<typeof Blob>[0]>[number]], {
        type: options.contentType ?? 'application/octet-stream',
      });
    formData.append('file', audio, filename);
    if (options.language) formData.append('language', options.language);
    if (options.model) formData.append('model', options.model);
    if (options.responseFormat) formData.append('response_format', options.responseFormat);
    if (options.prompt) formData.append('prompt', options.prompt);

    const response = await requestWithRetry({
      method: 'POST',
      url: `${this.http.base}/voice/transcribe`,
      headers: { Authorization: `Bearer ${this.http.credential}` },
      body: formData,
      rawBody: true,
      signal: options.signal,
    });
    if (response.status >= 400) throw await responseAPIError(response, 'POST');
    const contentType = response.headers.get('content-type') ?? '';
    if (contentType.includes('application/json')) {
      const data = await response.json() as Record<string, unknown>;
      return { ...data, text: String(data.text ?? data.transcript ?? '') };
    }
    return { text: await response.text() };
  }

  /**
   * Create a streaming VoiceSession over /ws/voice (call open() or use ttsStream()).
   */
  connect(options?: { timeoutMs?: number }): VoiceSession {
    return new VoiceSession({
      wsUrl: getAgentsWsUrlFromProductBase(this.http.base),
      credential: this.http.credential,
      timeoutMs: options?.timeoutMs,
    });
  }

  connectTranscription(options?: TranscriptionStartOptions & { timeoutMs?: number }): VoiceTranscriptionSession {
    return new VoiceTranscriptionSession({
      wsUrl: getAgentsWsUrlFromProductBase(this.http.base),
      credential: this.http.credential,
      timeoutMs: options?.timeoutMs,
      language: options?.language,
      model: options?.model,
      responseFormat: options?.responseFormat,
      prompt: options?.prompt,
    });
  }

  async *transcribeStream(
    options: TranscribeStreamOptions,
  ): AsyncGenerator<VoiceTranscriptionEvent, void, undefined> {
    const session = this.connectTranscription({
      timeoutMs: options.timeoutMs,
      language: options.language,
      model: options.model,
      responseFormat: options.responseFormat,
      prompt: options.prompt,
    });
    await session.open();
    try {
      yield* session.transcribe(options.audio, {
        base64: options.base64,
      });
    } finally {
      session.close();
    }
  }

  /**
   * One-shot streaming TTS: opens a session, speaks, closes.
   */
  async *ttsStream(options: TTSOptions & { timeoutMs?: number }): AsyncGenerator<VoiceChunkEvent, void, undefined> {
    const session = this.connect({ timeoutMs: options.timeoutMs });
    await session.open();
    try {
      yield* session.speak({
        text: options.text,
        voice: options.voice,
        language: options.language,
        format: options.responseFormat,
        chunks: true,
      });
    } finally {
      session.close();
    }
  }

  /**
   * One-shot streaming voice clone: opens a session, speaks, closes.
   */
  async *cloneStream(options: CloneOptions & { timeoutMs?: number }): AsyncGenerator<VoiceChunkEvent, void, undefined> {
    const session = this.connect({ timeoutMs: options.timeoutMs });
    await session.open();
    try {
      yield* session.speakClone({
        text: options.text,
        refAudio: options.refAudio,
        language: options.language,
        xVectorOnly: options.xVectorOnly,
        format: options.responseFormat,
        chunks: true,
      });
    } finally {
      session.close();
    }
  }

  /**
   * One-shot streaming voice design: opens a session, speaks, closes.
   */
  async *designStream(options: DesignOptions & { timeoutMs?: number }): AsyncGenerator<VoiceChunkEvent, void, undefined> {
    const session = this.connect({ timeoutMs: options.timeoutMs });
    await session.open();
    try {
      yield* session.speakDesign({
        text: options.text,
        description: options.description,
        language: options.language,
        format: options.responseFormat,
        chunks: true,
      });
    } finally {
      session.close();
    }
  }
}

File: sdk/hypercli/voice.py
from __future__ import annotations

"""Voice API client."""
import os
from pathlib import Path
from typing import TYPE_CHECKING, AsyncIterator
import base64

if TYPE_CHECKING:
    from .http import HTTPClient
    from .voice_stream import VoiceChunk, VoiceSession


def _encode_reference_audio(ref_audio: bytes | str | Path) -> str:
    if isinstance(ref_audio, bytes):
        audio_bytes = ref_audio
    else:
        audio_bytes = Path(ref_audio).read_bytes()
    return base64.b64encode(audio_bytes).decode()


def _default_voice_timeout() -> float:
    raw_value = os.environ.get("HYPER_VOICE_TIMEOUT_SECONDS", "300").strip()
    try:
        return float(raw_value)
    except ValueError:
        return 300.0


def _resolve_voice_timeout(timeout: float | None) -> float:
    if timeout is not None:
        return float(timeout)
    return _default_voice_timeout()


class VoiceAPI:
    """Voice capability API wrapper."""
    DEFAULT_TIMEOUT = 300.0

    def __init__(self, http: "HTTPClient"):
        self._http = http

    def tts(
        self,
        text: str,
        *,
        voice: str = "serena",
        language: str = "auto",
        response_format: str = "mp3",
        timeout: float | None = None,
    ) -> bytes:
        return self._http.post_bytes(
            "/agents/voice/tts",
            json={
                "text": text,
                "voice": voice,
                "language": language,
                "response_format": response_format,
            },
            timeout=_resolve_voice_timeout(timeout),
        )

    def clone(
        self,
        text: str,
        *,
        ref_audio: bytes | str | Path,
        language: str = "auto",
        x_vector_only: bool = True,
        response_format: str = "mp3",
        timeout: float | None = None,
    ) -> bytes:
        return self._http.post_bytes(
            "/agents/voice/clone",
            json={
                "text": text,
                "ref_audio_base64": _encode_reference_audio(ref_audio),
                "language": language,
                "x_vector_only": x_vector_only,
                "response_format": response_format,
            },
            timeout=_resolve_voice_timeout(timeout),
        )

    def design(
        self,
        text: str,
        *,
        description: str,
        language: str = "auto",
        response_format: str = "mp3",
        timeout: float | None = None,
    ) -> bytes:
        return self._http.post_bytes(
            "/agents/voice/design",
            json={
                "text": text,
                "instruct": description,
                "language": language,
                "response_format": response_format,
            },
            timeout=_resolve_voice_timeout(timeout),
        )

    def connect(self, *, timeout: float | None = None) -> "VoiceSession":
        """Create a streaming VoiceSession (open with 'async with' or open()).

        The session talks to /ws/voice and receives ordered audio chunks
        as the server renders them.
        """
        from .config import get_agents_ws_url_from_product_base
        from .voice_stream import VoiceSession

        ws_url = get_agents_ws_url_from_product_base(self._http.base_url)
        return VoiceSession(
            ws_url,
            self._http.api_key,
            timeout=_resolve_voice_timeout(timeout),
        )

    async def tts_stream(
        self,
        text: str,
        *,
        voice: str = "serena",
        language: str = "auto",
        response_format: str = "mp3",
        timeout: float | None = None,
    ) -> AsyncIterator["VoiceChunk"]:
        """One-shot streaming TTS: opens a session, speaks, closes."""
        async with self.connect(timeout=timeout) as session:
            async for chunk in session.speak(
                text,
                voice=voice,
                language=language,
                response_format=response_format,
                chunks=True,
            ):
                yield chunk
